ldms/src/sampler/llnl/lustre_client.c

```diff
@@ -0,0 +1,360 @@
+ * Produced at the Lawrence Livermore National Laboratory. Written by
+ * Christopher J. Morrone <morrone2@llnl.gov>
+ * All rights reserved.
+ *
+ * This work was performed under the auspices of the U.S. Department of
+ * Energy by Lawrence Livermore National Laboratory under
+ * Contract DE-AC52-07NA27344.
+ *
+ * This software is available to you under a choice of one of two
+ * licenses.  You may choose to be licensed under the terms of the GNU
+ * General Public License (GPL) Version 2, available from the file
+ * COPYING in the main directory of this source tree, or the BSD-type
+ * license below:
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ *
+ *      Redistributions of source code must retain the above copyright
+ *      notice, this list of conditions and the following disclaimer.
+ *
+ *      Redistributions in binary form must reproduce the above
+ *      copyright notice, this list of conditions and the following
+ *      disclaimer in the documentation and/or other materials provided
+ *      with the distribution.
+ *
+ *      Neither the name of Sandia nor the names of any contributors may
+ *      be used to endorse or promote products derived from this software
+ *      without specific prior written permission.
+ *
+ *      Neither the name of Open Grid Computing nor the names of any
+ *      contributors may be used to endorse or promote products derived
+ *      from this software without specific prior written permission.
+ *
+ *      Modified source versions must be plainly marked as such, and
+ *      must not be misrepresented as being the original software.
+ *
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+#include <limits.h>
+#include <string.h>
+#include <dirent.h>
+#include <coll/rbt.h>
+#include <sys/queue.h>
+#include <time.h>
+#include <unistd.h>
+#include "ldms.h"
+#include "ldmsd.h"
+#include "ldms_jobid.h"
+#include "lustre_client.h"
+#include "lustre_client_general.h"
+
+#define _GNU_SOURCE
+
+#define LLITE_PATH "/proc/fs/lustre/llite"
+#define OSD_SEARCH_PATH "/proc/fs/lustre"
+
+ldmsd_msg_log_f log_fn;
```

Comment:
this cannot be global. it needs to be passed to supporting library code via arg list if the sampler is split into library code.

Reply:
Why?  This is a plugin, which presumably is dlopen()ed by ldmsd, right?  So unless ldmsd dlsym()s the plugin's log_fn, it doesn't effect the larger program, right?  You say, into "library code".  This plugin is not split into library code, it is merely broken up into multiple files for organizational purposes.

Reply:
ldmsd is intended to be used in a variety of exotic situations, including fully static linking/embedded. So we need to keep the symbol scoping clean, which in this case is not at all hard: just pass log_fn to the init function in your support file and store it in a static variable of that file as well. pass with_jobid through the *..._sample function signature.

Reply:
Ah, I see.